File: python/3372.maximize-the-number-of-target-nodes-after-connecting-trees-i.py
class Solution:
    def buildAdjList(self, edges: list[list[int]]):
        n = len(edges) + 1
        adj = [[] for _ in range(n)]

        for u, v in edges:
            adj[u].append(v)
            adj[v].append(u)

        return adj

    def dfs(self, node: int, parent: int, depth: int, adj: list[list[int]]) -> int:
        if depth < 0:
            return 0

        count = 1
        for neighbor in adj[node]:
            if neighbor == parent:
                continue

            count += self.dfs(neighbor, node, depth - 1, adj)

        return count

    def maxTargetNodes(
        self, edges1: list[list[int]], edges2: list[list[int]], k: int
    ) -> list[int]:
        adj1 = self.buildAdjList(edges1)

        if k < 2:
            return [self.dfs(i, -1, k, adj1) + k for i in range(len(adj1))]

        adj2 = self.buildAdjList(edges2)
        max2 = max(self.dfs(i, -1, k - 1, adj2) for i in range(len(adj2)))

        return [self.dfs(i, -1, k, adj1) + max2 for i in range(len(adj1))]

File: golang/3372.maximize-the-number-of-target-nodes-after-connecting-trees-i.go
func maxTargetNodes(edges1 [][]int, edges2 [][]int, k int) []int {
	adj1 := buildAdjTree(edges1)

	if k < 2 {
		res := make([]int, len(adj1))
		for i := range res {
			res[i] = dfsDepthCount(adj1, i, -1, k) + k
		}
		return res
	}

	adj2 := buildAdjTree(edges2)
	max2 := 0

	for i := range adj2 {
		max2 = max(max2, dfsDepthCount(adj2, i, -1, k-1))
	}

	res := make([]int, len(adj1))
	for i := range res {
		res[i] = dfsDepthCount(adj1, i, -1, k) + max2
	}
	return res
}

func dfsDepthCount(adj [][]int, node, parent, depth int) int {
	if depth < 0 {
		return 0
	}
	count := 1
	for _, child := range adj[node] {
		if child == parent {
			continue
		}
		count += dfsDepthCount(adj, child, node, depth-1)
	}
	return count
}

func buildAdjTree(edges [][]int) [][]int {
	adj := make([][]int, len(edges)+1)
	for _, edge := range edges {
		adj[edge[0]] = append(adj[edge[0]], edge[1])
		adj[edge[1]] = append(adj[edge[1]], edge[0])
	}
	return adj
}
